core/shared/src/main/scala/zio/stm/TQueue.scala

```diff
@@ -54,9 +54,71 @@ final class TQueue[A] private (val capacity: Int, ref: TRef[ScalaQueue[A]]) {
       .map(_.splitAt(max))
       .flatMap(split => ref.set(split._2) *> STM.succeed(split._1))
       .map(_.toList)
+
+  /**
+   * View the next element in the queue without removing it. retries if the queue is empty.
+   */
+  def peek: STM[Nothing, A] =
+    ref.get.flatMap(
+      _.headOption match {
+        case Some(a) => STM.succeed(a)
+        case None    => STM.retry
+      }
+    )
+
+  /**
+   * View the last element inserted into the queue. retries if the queue is empty
+   */
+  def last: STM[Nothing, A] =
+    ref.get.flatMap(
+      _.lastOption match {
+        case Some(a) => STM.succeed(a)
+        case None    => STM.retry
+      }
+    )
+
+  /**
+   * Checks if the Queue is empty
+   */
+  def isEmpty: STM[Nothing, Boolean] = ref.get.map(_.isEmpty)
+
+  /**
+   * Checks if the Queue is at capacity
+   */
+  def isFull: STM[Nothing, Boolean] = ref.get.map(_.size == capacity)
+
+  /**
+   * Drops elements from the queue while they dont satisfy the predicate until it does and takes it.
+   */
+  def seek(f: A => Boolean): STM[Nothing, A] = {
+    @annotation.tailrec
+    def go(q: ScalaQueue[A]): STM[Nothing, A] =
+      q.dequeueOption match {
+        case Some((a, as)) =>
+          if (f(a)) ref.set(as) *> STM.succeed(a)
+          else go(as)
+        case None => STM.retry
+      }
+
+    ref.get.flatMap(go)
+  }
+
+  /**
+   * Compares two Queues with the same element type and returns the longest one. If equal it returns the left side.
+   */
+  def longest(t: TQueue[A]): STM[Nothing, TQueue[A]] =
+    (size <*> t.size).map { case (a, b) => if (a >= b) this else t }
+
+  /**
+   * Returns the underlying queue
+   */
+  def underlying: STM[Nothing, TRef[ScalaQueue[A]]] = STM.succeed(ref)
```

Comment:
I'm not sure about this one. While I understand the motivation, even when wrapped in `TRef` it feels like a leak. In case it ever happens that we need to change the internal encoding, this might be an ugly blocker.

Reply:
It's true. Maybe we should avoid it...